neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,194 @@
+    """
+
+    def __init__(self,
+                 channels,
+                 n_heads,
+                 head_n_channels,
+                 pos_dim,
+                 use_positional_encoding=True,    # use positional encoding
+                 project_query=True,
+                 ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
+        self.n_heads = n_heads
+        self.head_n_channels = head_n_channels
+
+        self.project_query = project_query
+        if project_query:
+            self.to_q = nn.Linear(channels, head_n_channels * n_heads, bias=False)
+        else:
+            self.to_q = nn.Identity()
+
+        self.to_k = nn.Linear(channels, head_n_channels * n_heads, bias=False)
+
+        self.k_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+        self.v_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+
+        self.to_v = nn.Linear(channels, head_n_channels * n_heads, bias=False)
+
+        self.to_out = nn.Linear(head_n_channels * n_heads, channels) if head_n_channels * n_heads != channels else nn.Identity()
+
+        self.use_positional_encoding = use_positional_encoding
+        self.pos_dim = pos_dim
+
+        self.init_gain = 1 / math.sqrt(head_n_channels)
+        self.diagonal_weight = self.init_gain
+        self.initialize_qkv_weights()
+
+    def init_weight(self, weight, inif_fn):
+        # initialization for the projection matrix
+        # basically initialize the weights for each n_heads
+        # to add the diagonal bias, it requires input dim = head_n_channels
+        # W = init_fn(W) + I * diagonal_weight
+        # init_fn is usually xavier_uniform_
+
+        for param in weight.parameters():
+            if param.ndim > 1:
+                for h in range(self.n_heads):
+                    inif_fn(param[h * self.head_n_channels:(h + 1) * self.head_n_channels, :], gain=self.init_gain)
+                    if self.head_n_channels == self.channels:
+                        diagonal_bias = self.diagonal_weight * torch.diag(torch.ones(param.size(-1), dtype=torch.float32))
+                        param.data[h * self.head_n_channels:(h + 1) * self.head_n_channels, :] += diagonal_bias
+
+    def initialize_qkv_weights(self):
+        init_fn = xavier_uniform_
+
+        if self.project_query:
+            self.init_weight(self.to_q, init_fn)
+        self.init_weight(self.to_k, init_fn)
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, u, norm_fn):
+        # u: the input or transformed function
+        batch_size = u.shape[0]
+        u = u.view(batch_size*self.n_heads, -1, self.head_n_channels)
+        u = norm_fn(u)    # layer norm with channel dimension or instance norm with spatial dimension
+        return u.view(batch_size, self.n_heads, -1, self.head_n_channels)
+
+    def forward(self,
+                u_x,
+                pos_x,
+                pos_emb=None,    # positional encoding module for encoding q/k
```

Comment:
Can we spell out the variables, e.g. here pos_embedding or even positional_embedding? It's a little longer to type but makes code very readable

Reply:
I named it positional_embedding_module, if it's too lengthy I can also change it to positional_embedding or sth else